### ISSUE TITLE
VS code extension doesn't detect lsp binary

### ISSUE THREAD

**Issue by WannesMalfait** [nvarner/typst-lsp#233]
- **Component**: 
    - VSCode Extension
- **Extension version**: 9.0.0
- **OS version and name**: PopOS! 22.04 LTS 

- I am on the [latest](https://github.com/nvarner/typst-lsp/tags) stable version of the extension/LSP.
- I have searched the [issues](https://github.com/nvarner/typst-lsp/issues) of this repo and believe that this is not a duplicate.

## Issue
I just upgraded the extension from 0.7.2 to 9.0.0 and wanted to try it out. When I opened a `*.typ` file I got the following error:
`Failed to activate typst-lsp: Error: Could not find a valid typst-lsp binary bundled or in PATH`.
I tried restarting vs code or re-installing the extension but the same error appears.

Does this mean I have to install the LSP separately now? I didn't have to do this on previous versions.

### Logs
No logs seem to be visible.

### Possible cause
Possibly caused by #220 (since the error message was added there)?

**Comment by nvarner** [comment]
It should still be bundled. What architecture is your computer? Have you modified any of the configuration?

**Comment by WannesMalfait** [comment]
No, I did not change any of the settings. Here is my neofetch:
![Screenshot from 2023-08-11 20-36-07](https://github.com/nvarner/typst-lsp/assets/46323945/ab2c62cc-52e3-4d8b-8493-127411eaad6d)

**Comment by nvarner** [comment]
I'm working on #232 right now, and I'd like to fix this once I'm done with that. Unfortunately, nothing stands out as a likely cause, so I may not be able to make much progress. In the meantime, [downloading the binary as `typst-lsp`](https://github.com/nvarner/typst-lsp/releases/tag/v0.9.0), then either putting it in your path or adding a path to it in the settings, should work around the issue for now.

If anyone else is having this same issue, please post here with:
- your operating system
- your processor architecture (likely x86_64 or ARM; if you're not sure, I can find it from your CPU or computer model)
- which LSP client you're using (if you're not sure, probably the VS Code extension)
- any settings you may have changed
A larger sample size can help a lot.

**Comment by nvarner** [comment]
Since you probably know the relevant code best @fschwalbe, any guesses what could cause this? I redid some CI yesterday after things failed to compile, so my best thought is that one of my changes broke something (but if so, no idea what broke).

**Comment by WannesMalfait** [comment]
I investigated a little bit and found this:
https://github.com/nvarner/typst-lsp/blob/4b367b8773c643c9cc198a4b6f9fb3abe837372d/editors/vscode/src/extension.ts#L87 seems to be returning `127` as status instead of `0 `. According to [this](https://tldp.org/LDP/abs/html/exitcodes.html) that exit code means "command not found".

**Comment by WannesMalfait** [comment]
I tried running the binary bundled with the extension manually and got this:
```
> /home/wannes/.vscode/extensions/nvarner.typst-lsp-0.9.0-linux-x64/out/typst-lsp

/home/wannes/.vscode/extensions/nvarner.typst-lsp-0.9.0-linux-x64/out/typst-lsp: error while loading shared libraries: libssl.so.1.1: cannot open shared object file: No such file or directory
```
The same error occurs in version `0.8.1`, so that seems to be the problem
Running the binary manually for version `0.7.2` just runs normally and waits for lsp commands.

**Comment by nvarner** [comment]
That would do it. If you're not familiar, that's happening because you don't have OpenSSL installed. (Is that intentional, or just not installed on your system? If it isn't installed by default, that will affect how I approach this.)

I'm not yet sure the best solution, but I can fix this now; it will probably come along with #234.

**Comment by WannesMalfait** [comment]
OpenSSL was already installed on my system (and it is by default I think).
```
> openssl version
OpenSSL 3.1.1 30 May 2023 (Library: OpenSSL 3.1.1 30 May 2023)
> sudo apt install libssl-dev
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
libssl-dev is already the newest version (3.0.2-0ubuntu1.10).
0 upgraded, 0 newly installed, 0 to remove and 0 not upgraded.
> sudo apt install openssl
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
openssl is already the newest version (3.0.2-0ubuntu1.10).
0 upgraded, 0 newly installed, 0 to remove and 0 not upgraded.
```

**Comment by nvarner** [comment]
That is not what I expected the problem to be! But that helps a lot to know (as has the rest of your investigation; thank you for figuring all this out).

If you're curious, my best guess is that either `reqwest` is pulling an old version of OpenSSL, or the GitHub actions runner I was using gave it an old version. It looks like OpenSSL 1.1.1 is a bit old and is nearly at EOL, so that definitely has to be changed. If I can't get it to link to OpenSSL 3.x, I'll either have it build with a bundled version of OpenSSL, or default to rustls.

**Comment by WannesMalfait** [comment]
BTW, don't know if this helps, but I cloned the repository, built the lsp with `cargo build` then pointed the extension settings server path to that binary, and that worked (using that for now). So it seems indeed that the pre-built binary is linked against an older version of OpenSSL

**Comment by nvarner** [comment]
The release binaries were built using an Ubuntu 20.x image, so I'd bet that's the problem. Again, thanks for all the help! I'll hopefully be able to get a fix in soon.

**Comment by Myriad-Dreamin** [comment]
> The release binaries were built using an Ubuntu 20.x image, so I'd bet that's the problem. Again, thanks for all the help! I'll hopefully be able to get a fix in soon.

What about using rustls for reqwest whenever possible? I can confirm that only riscv and windows arm targets force a native-tls, because the ring crate fails in those targets.

**Comment by nvarner** [comment]
I've written a fix for this that should show up in the v0.9.1 release being built right now. Rustls is now the default wherever supported. There is also new error reporting in the VS Code extension to make debugging something like this easier in the future.
